### PR TITLE
IGNITE-19228 Schema validation during tx processing: common framework

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-19228

### CODE REVIEW

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/SchemaCompatValidator.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+            boolean commit,
+            @Nullable HybridTimestamp commitTimestamp
+    ) {
+        if (!commit) {
+            return completedFuture(ForwardValidationResult.success());
+        }
+
+        HybridTimestamp beginTimestamp = TransactionIds.beginTimestamp(txId);
+
+        List<UUID> tableIds = enlistedGroupIds.stream()
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/SchemaCompatValidator.java:69]: I think this intermediate variable is redundant, you can use the stream directly inside `thenApply` below

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+    @Override
+    public List<FullTableSchema> tableSchemaVersionsBetween(UUID tableId, HybridTimestamp fromIncluding, HybridTimestamp toIncluding) {
+        SchemaRegistry schemaRegistry = schemaManager.schemaRegistry(tableId);
+
+        SchemaDescriptor schemaDescriptor = schemaRegistry.schema();
+        FullTableSchema fullSchema = new FullTableSchema(
+                1,
+                1,
+                schemaDescriptor.columnNames().stream()
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java:72]: Please extract this into a variable

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        return getColumnType(type.spec());
+    }
+
+    /**
+     * Detects {@link ColumnType} by {@link NativeTypeSpec}.
+     *
+     * @param spec Native type spec.
+     * @return Detected {@link ColumnType}.
+     */
+    public static ColumnType getColumnType(NativeTypeSpec spec) {
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java:106]: There exists an exact copy of this method in `ClientTableCommon`, I would suggest to extract it into a common place

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+                return ColumnType.TIME;
+
+            case DATETIME:
+                return ColumnType.DATETIME;
+
+            case TIMESTAMP:
+                return ColumnType.TIMESTAMP;
+
+            default:
+                throw new IgniteException(PROTOCOL_ERR, "Unsupported native type: " + spec);
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java:157]: Wait a second, you indeed copy-pasted it from `ClientTableCommon`, shame on you!

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    private static TableColumnDescriptor columnType(String colName, Column column) {
+        return new TableColumnDescriptor(
+                colName,
+                columnTypeFromNativeType(column.type()),
+                column.nullable(),
+                DefaultValue.constant(column.defaultValue())
+        );
+    }
+
+    private static ColumnType columnTypeFromNativeType(NativeType type) {
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java:96]: This method seems redundant to me

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+                            return columnType(colName, column);
+                        })
+                        .collect(toList()),
+                List.of()
+        );
+
+        return List.of(fullSchema);
+    }
+
+    private static TableColumnDescriptor columnType(String colName, Column column) {
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java:87]: Should be called `columnDescriptor`

Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java:87]: `colName` parameter is redundant, `column` already contains a name

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        return new TableColumnDescriptor(
+                colName,
+                columnTypeFromNativeType(column.type()),
+                column.nullable(),
+                DefaultValue.constant(column.defaultValue())
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java:92]: Are sure this is correct? There are multiple types of default values

File: modules/table/src/test/java/org/apache/ignite/internal/table/distributed/replication/PartitionReplicaListenerTest.java [hunk 8]
```diff
@@ -290,6 +314,8 @@ public void beforeTest(@InjectConfiguration DataStorageConfiguration dsCfg) {
 
         lenient().when(safeTimeClock.waitFor(any())).thenReturn(completedFuture(null));
 
+        lenient().when(schemas.waitForSchemasAvailability(any())).thenReturn(completedFuture(null));
```
Comment (sashapolo) [inline modules/table/src/test/java/org/apache/ignite/internal/table/distributed/replication/PartitionReplicaListenerTest.java:317]: I know that it's not related to this PR, but all these `lenient()` calls can be removed using the `MockitoSettings` annotation

File: modules/table/src/testFixtures/java/org/apache/ignite/internal/table/impl/DummySchemas.java [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+        return completedFuture(null);
+    }
+
+    @Override
+    public List<FullTableSchema> tableSchemaVersionsBetween(UUID tableId, HybridTimestamp fromIncluding, HybridTimestamp toIncluding) {
+        SchemaDescriptor schemaDescriptor = schemaRegistry.schema();
+        FullTableSchema fullSchema = new FullTableSchema(
+                1,
+                1,
+                schemaDescriptor.columnNames().stream()
```
Comment (sashapolo) [inline modules/table/src/testFixtures/java/org/apache/ignite/internal/table/impl/DummySchemas.java:59]: Please extract this into a variable

File: modules/table/src/testFixtures/java/org/apache/ignite/internal/table/impl/DummySchemas.java [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+        return new TableColumnDescriptor(
+                colName,
+                columnTypeFromNativeType(column.type()),
+                column.nullable(),
+                DefaultValue.constant(column.defaultValue())
+        );
+    }
+
+    private static ColumnType columnTypeFromNativeType(NativeType type) {
+        return NonHistoricSchemas.getColumnType(type.spec());
```
Comment (sashapolo) [inline modules/table/src/testFixtures/java/org/apache/ignite/internal/table/impl/DummySchemas.java:84]: Why do you reuse this method from `NonHistoricSchemas` and not the method that creates a `TableColumnDescriptor` from a `Column`?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 11]
```diff
@@ -976,13 +983,34 @@ private CompletableFuture<Void> processTxFinishAction(TxFinishReplicaRequest req
         List<TablePartitionId> aggregatedGroupIds = request.groups().values().stream()
                 .flatMap(List::stream)
                 .map(IgniteBiTuple::get1)
-                .collect(Collectors.toList());
+                .collect(toList());
 
         UUID txId = request.txId();
 
-        boolean commit = request.commit();
+        return schemaCompatValidator.validateForwards(txId, aggregatedGroupIds, request.commit(), request.commitTimestamp())
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:990]: I see two alternatives to the current approach.

One is to branch on `request.commit()` in the `PartitionReplicaListener` and, if it's a commit, do what is done there now (validate, then invoke `finishAndCleanup()`, then maybe throw); if it's an abort, just call `finishAndCleanup()`. But the result is that `finishAndCleanup()` is called twice in different branches, and this seems weird.

Another alternative is to build a method like `validateForwardsIfCommit()`, but it would still produce a future of `ForwardValidationResult`, so we would have to return `success()` if it's an abort. Still looks weird: we kinda did not validate, but we returned a validation result. We could also add a third state to the validation result, like 'no need for validation', but it would work exactly like 'success', so this also looks weird.

What changes do you envision here?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 12]
```diff
@@ -976,13 +983,34 @@ private CompletableFuture<Void> processTxFinishAction(TxFinishReplicaRequest req
...
+                .thenCompose(validationResult -> {
+                    boolean stillCommit = request.commit() && validationResult.isOk();
 
-        CompletableFuture<Object> changeStateFuture = finishTransaction(aggregatedGroupIds, txId, commit);
+                    return finishAndCleanup(request, stillCommit, aggregatedGroupIds, txId)
+                            .thenCompose(result -> {
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:995]: I return a future to signal an exception, so `thenAccept()` does not fit. Do you suggest throwing the exception?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java [hunk 13]
```diff
@@ -976,13 +983,34 @@ private CompletableFuture<Void> processTxFinishAction(TxFinishReplicaRequest req
...
+                                if (validationResult.isOk()) {
+                                    return completedFuture(result);
+                                } else {
+                                    return failedFuture(new AbortDueToIncompatibleSchemaException("Commit failed because schema "
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:999]: The idea is that, if validation fails, we change the intention from commit to abort, so finishTransaction() will actually abort the transaction, and only then will we signal the exception

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/SchemaCompatValidator.java (hunk 1 above)
Comment (rpuch) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/SchemaCompatValidator.java:69]: It looks too cumbersome inside `for`. Introducing a variable allows to easier read the `for` loop, imho.

File: modules/table/src/test/java/org/apache/ignite/internal/table/distributed/replication/PartitionReplicaListenerTest.java (hunk 8 above)
Comment (rpuch) [reply modules/table/src/test/java/org/apache/ignite/internal/table/distributed/replication/PartitionReplicaListenerTest.java:317]: I'm not sure it's a good thing to just make all stubbings lenient. I would prefer to be explicit.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java (hunk 4 above)
Comment (rpuch) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java]: I'm sorry :(

File: modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/asserts/CompletableFutureAssert.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+/**
+ * Assertions related to {@link CompletableFuture}.
+ */
+public class CompletableFutureAssert {
```
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/asserts/CompletableFutureAssert.java:31]: Can we add a `Matcher` and an overload that allows us to inspect the exception? This way we can write something like `assertThat(f, willThrow(e -> ...)`

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java]: yes

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java (hunk 7 above)
Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java:95]: Because looks like there is also a functional default value type, will it work in this case?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/RebalanceRaftGroupEventsListener.java [hunk 15]
```diff
@@ -347,12 +347,12 @@ private void doOnNewPeersConfigurationApplied(PeersAndLearners configuration) {
             // For further addition
             Set<Assignment> calculatedSwitchAppend = union(retrievedSwitchAppend, reducedNodes);
             calculatedSwitchAppend = subtract(calculatedSwitchAppend, addedNodes);
-            calculatedSwitchAppend = intersect(calculatedAssignments, calculatedSwitchAppend);
+            calculatedSwitchAppend = CollectionUtils.intersect(calculatedAssignments, calculatedSwitchAppend);
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/RebalanceRaftGroupEventsListener.java:350]: I propose to add a static import here, looks strange otherwise

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/FullTableSchema.java [hunk 16]
```diff
@@ -125,34 +117,24 @@ public TableDefinitionDiff diffFrom(FullTableSchema prevSchema) {
...
+    private static <T> Map<String, T> toMapByName(List<T> elements, Function<T, String> nameExtractor) {
+        return elements.stream().collect(toMap(nameExtractor, identity()));
     }
 
-    private static Set<String> intersect(Set<String> set1, Set<String> set2) {
-        Set<String> result = new HashSet<>(set1);
-        result.retainAll(set2);
-        return result;
+    private static <T> List<T> subtractKeyed(Map<String, T> diminuend, Map<String, T> subtrahend) {
+        return diminuend.keySet().stream()
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/FullTableSchema.java:134]: You can use `entrySet` here to avoid calling `get` afterwards

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/FullTableSchema.java [hunk 17]
```diff
@@ -125,34 +117,24 @@ public TableDefinitionDiff diffFrom(FullTableSchema prevSchema) {
...
-        return result;
+    private static <T> Map<String, T> toMapByName(List<T> elements, Function<T, String> nameExtractor) {
+        return elements.stream().collect(toMap(nameExtractor, identity()));
     }
 
-    private static Set<String> intersect(Set<String> set1, Set<String> set2) {
-        Set<String> result = new HashSet<>(set1);
-        result.retainAll(set2);
-        return result;
+    private static <T> List<T> subtractKeyed(Map<String, T> diminuend, Map<String, T> subtrahend) {
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/FullTableSchema.java:133]: `diminuend`? What does that mean?

File: modules/table/src/main/java/org/apache/ignite/internal/utils/RebalanceUtil.java [hunk 18]
```diff
@@ -468,7 +468,7 @@ public static CompletableFuture<Void> handleReduceChanged(MetaStorageManager met
      * @return Result of the subtraction.
      */
     public static <T> Set<T> subtract(Set<T> minuend, Set<T> subtrahend) {
```
Comment (sashapolo) [inline modules/table/src/main/java/org/apache/ignite/internal/utils/RebalanceUtil.java:470]: Can we get rid of this method?

File: modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java [hunk 19]
```diff
@@ -596,4 +597,15 @@ public static <T> Set<T> difference(@Nullable Set<T> a, @Nullable Set<T> b) {
...
     }
+
+    /**
+     * Returns an intersection of two sets.
+     *
+     * @param op1 First operand.
+     * @param op2 Second operand.
+     * @return Result of the intersection.
+     */
+    public static <T> Set<T> intersect(Set<T> op1, Set<T> op2) {
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/util/CollectionUtils.java:608]: No subtrahend =(

File: modules/core/src/test/java/org/apache/ignite/internal/util/CollectionUtilsTest.java [hunk 20]
```diff
@@ -359,4 +361,9 @@ private void testSetOf(Collection<Integer> data) {
         assertThrows(UnsupportedOperationException.class, () -> copy.add(42));
         assertThrows(UnsupportedOperationException.class, () -> copy.remove(3));
     }
+
+    @Test
+    void testIntersect() {
```
Comment (sashapolo) [inline modules/core/src/test/java/org/apache/ignite/internal/util/CollectionUtilsTest.java:366]: Let's add a test for en empty intersection as well

File: modules/table/src/test/java/org/apache/ignite/internal/table/distributed/replication/PartitionReplicaListenerTest.java (hunk 8 above)
Comment (sashapolo) [reply modules/table/src/test/java/org/apache/ignite/internal/table/distributed/replication/PartitionReplicaListenerTest.java:317]: ok

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java:990]: > But the result is that finishAndCleanup() is called twice in different branches, and this seems weird.

I don't know but it looks perfectly fine to me:

```
if (request.commit()) {
    return schemaCompatValidator.validateForwards(txId, aggregatedGroupIds, request.commitTimestamp())
            .thenCompose(validationResult -> {
                return finishAndCleanup(request, validationResult.isSuccessful(), aggregatedGroupIds, txId)
                        .thenCompose(result -> failureIfValidationFailed(validationResult));
            });
} else {
    return finishAndCleanup(request, false, aggregatedGroupIds, txId);
}
```

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java]: oops, I missed the `stillCommit` part

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/SchemaCompatValidator.java (hunk 1 above)
Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/SchemaCompatValidator.java:69]: I think it looks ok if extracted into a method:
```
ForwardValidationResult validateTablesSchemas(
        List<TablePartitionId> enlistedGroupIds,
        HybridTimestamp beginTimestamp,
        HybridTimestamp commitTimestamp
) {
    return enlistedGroupIds.stream()
            .map(TablePartitionId::tableId)
            .distinct()
            .map(tableId -> validateSchemaCompatibility(beginTimestamp, commitTimestamp, tableId))
            .filter(validationResult -> !validationResult.isSuccessful())
            .findAny()
            .orElse(ForwardValidationResult.success());
}
```

Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/SchemaCompatValidator.java:69]: Or, if you still like your `for` cycle, I think it's better to use a `Set` instead of `distinct().toList()`

File: modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/asserts/CompletableFutureAssert.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/asserts/CompletableFutureAssert.java:31]: A matcher gives a binary answer (matches/does not match) and a description of what does not match. This is a pretty rigid format if you want to make a few assertions. On the other hand, if we just add assertions (throwing an AssertionError) 'inside' a matcher, this looks like an abuse of the idea of a matcher.

Each tool is good for its situation. Sometimes a matcher is the right tool, sometimes direct inspection is better. It seems correct to have both tools and use them accordingly. Moreover, this does not bloat the code too much, the class we are discussing is rather small and simple.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java (hunk 7 above)
Comment (rpuch) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java:96]: This is a temporary code that will be thrown away when we finish the switch. The column definition that is produced is actually 'fake', it's not used anywhere because this class is only used for schema validation, and for now it always returns just one schema, so no actual column validation happens.

Also, it's impossible to directly convert `Column`'s functional default value provider (which is a `DefaultValueGenerator`) to `TableColumnDefinition`'s functional default value (which is a string - just a function name).

I amended the comment to reflect that the conversion is approximate.

File: modules/table/src/main/java/org/apache/ignite/internal/utils/RebalanceUtil.java (hunk 18 above)
Comment (rpuch) [reply modules/table/src/main/java/org/apache/ignite/internal/utils/RebalanceUtil.java:470]: Not in this PR probably

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java (hunk 7 above)
Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/schema/NonHistoricSchemas.java:96]: got it

File: modules/table/src/main/java/org/apache/ignite/internal/utils/RebalanceUtil.java (hunk 18 above)
Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/utils/RebalanceUtil.java:470]: why? Is it used a lot?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/SchemaCompatValidator.java (hunk 1 above)
Comment (rpuch) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/SchemaCompatValidator.java]: Comparing the two options, I still prefer the `for` loop as the one with the stream makes you think a bit harder to understand it. But I like the ideas of using `Set` and extracting the method.

File: modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/asserts/CompletableFutureAssert.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/core/src/testFixtures/java/org/apache/ignite/internal/testframework/asserts/CompletableFutureAssert.java:31]: ok, my suggestion was a little bit too eager and does not cover your case, I agree. I just don't like this logic duplication between two classes, but ok, let's keep this class(

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/SchemaCompatValidator.java (hunk 1 above)
Comment (sashapolo) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/SchemaCompatValidator.java]: ok

File: modules/table/src/main/java/org/apache/ignite/internal/utils/RebalanceUtil.java (hunk 18 above)
Comment (rpuch) [reply modules/table/src/main/java/org/apache/ignite/internal/utils/RebalanceUtil.java]: You are right, it's not a big deal. Removed it.

**Review by sashapolo — APPROVED** [review]
